Exception should be direct cause of `AssertionError` from `raises` context manager on failure for exception to match pattern rather than context

#### What's the problem this feature will solve?

Currently, the `AssertionError: Regex pattern did not match` thrown on exit of a `pytest.raises` context when an exception was raised but the provided pattern was not matched is attached to the exception actually raised as its `__context__` attribute implicitly. This causes the test failure report to incorrectly claim that 'During handling of the above exception, another exception occurred'.

#### Describe the solution you'd like

It should be attached to the `__cause__` attribute instead, such that the connective would be 'The above exception was the direct cause of the following exception', which is more correct. Conceptually, that would mean doing something along the lines of the following:

```python
raise AssertionError('Regex pattern did not match') from exc
```

where `exc` is the exception having occurred. However, I'm not sure where the logic is; I think it's in the `matches` method of the `RaisesExc` class.

--- a/AUTHORS
+++ b/AUTHORS
@@ -180,6 +180,7 @@ Fraser Stark
 Freya Bruhin
 Gabriel Landau
 Gabriel Reis
+Garion Milazzo
 Garvit Shubham
 Gene Wood
 George Kussumoto

--- a/changelog/14389.bugfix.rst
+++ b/changelog/14389.bugfix.rst
@@ -1,0 +1,1 @@
+Improved ``pytest.raises(..., match=...)`` failures to report the mismatched exception as the direct cause of the resulting ``AssertionError``.

--- a/src/_pytest/raises.py
+++ b/src/_pytest/raises.py
@@ -698,7 +698,7 @@ class RaisesExc(AbstractRaises[BaseExcT_co_default]):
         if not self.matches(exc_val):
             if self._just_propagate:
                 return False
-            raise AssertionError(self._fail_reason) from exc_val
+            raise AssertionError(self._fail_reason) from None
 
         # Cast to narrow the exception type now that it's verified....
         # even though the TypeGuard in self.matches should be narrowing

--- a/src/_pytest/raises.py
+++ b/src/_pytest/raises.py
@@ -698,7 +698,7 @@ class RaisesExc(AbstractRaises[BaseExcT_co_default]):
         if not self.matches(exc_val):
             if self._just_propagate:
                 return False
-            raise AssertionError(self._fail_reason)
+            raise AssertionError(self._fail_reason) from exc_val
 
         # Cast to narrow the exception type now that it's verified....
         # even though the TypeGuard in self.matches should be narrowing

--- a/testing/python/raises.py
+++ b/testing/python/raises.py
@@ -179,9 +179,7 @@ class TestRaises:
         result.stdout.no_fnmatch_line("*File*")
         result.stdout.no_fnmatch_line("*line*")
 
-    def test_raises_match_failure_uses_direct_cause(
-        self, pytester: Pytester
-    ) -> None:
+    def test_raises_match_failure_uses_direct_cause(self, pytester: Pytester) -> None:
         pytester.makepyfile(
             """
             import pytest

--- a/testing/python/raises.py
+++ b/testing/python/raises.py
@@ -179,6 +179,31 @@ class TestRaises:
         result.stdout.no_fnmatch_line("*File*")
         result.stdout.no_fnmatch_line("*line*")
 
+    def test_raises_match_failure_uses_direct_cause(
+        self, pytester: Pytester
+    ) -> None:
+        pytester.makepyfile(
+            """
+            import pytest
+
+            def test_raises_match_failure():
+                with pytest.raises(ValueError, match="expected"):
+                    raise ValueError("actual")
+            """
+        )
+        result = pytester.runpytest("--tb=short")
+        assert result.ret == 1
+        result.stdout.fnmatch_lines(
+            [
+                "*ValueError: actual",
+                "*The above exception was the direct cause of the following exception:*",
+                "*E*AssertionError: Regex pattern did not match.*",
+            ]
+        )
+        result.stdout.no_fnmatch_line(
+            "*During handling of the above exception, another exception occurred:*"
+        )
+
     def test_noclass(self) -> None:
         with pytest.raises(TypeError):
             with pytest.raises("wrong"):  # type: ignore[call-overload]

--- a/testing/python/raises.py
+++ b/testing/python/raises.py
@@ -179,7 +179,9 @@ class TestRaises:
         result.stdout.no_fnmatch_line("*File*")
         result.stdout.no_fnmatch_line("*line*")
 
-    def test_raises_match_failure_uses_direct_cause(self, pytester: Pytester) -> None:
+    def test_raises_match_failure_suppresses_exception_context(
+        self, pytester: Pytester
+    ) -> None:
         pytester.makepyfile(
             """
             import pytest
@@ -193,10 +195,12 @@ class TestRaises:
         assert result.ret == 1
         result.stdout.fnmatch_lines(
             [
-                "*ValueError: actual",
-                "*The above exception was the direct cause of the following exception:*",
                 "*E*AssertionError: Regex pattern did not match.*",
             ]
+        )
+        result.stdout.no_fnmatch_line("*ValueError: actual")
+        result.stdout.no_fnmatch_line(
+            "*The above exception was the direct cause of the following exception:*"
         )
         result.stdout.no_fnmatch_line(
             "*During handling of the above exception, another exception occurred:*"

